src/mod/applications/mod_http_cache/mod_http_cache.c

```diff
@@ -1185,6 +1331,175 @@ static switch_status_t http_get(url_cache_t *cache, http_profile_t *profile, cac
+	switch_CURLcode curl_status = CURLE_UNKNOWN_OPTION;
+
+	/* set up HTTP HEAD */
+	exists_data.fd = 0;
+	exists_data.url = url;
+
+	/* find profile for domain */
+	if (!profile) {
+		profile = url_cache_http_profile_find_by_fqdn(cache, url->url);
+	}
+
+	if (profile && profile->append_headers_ptr) {
+		headers = profile->append_headers_ptr(profile, headers, "HEAD", 0, "", url->url, 0, NULL);
+	}
+
+	curl_handle = switch_curl_easy_init();
+	switch_log_printf(SWITCH_CHANNEL_SESSION_LOG(session), SWITCH_LOG_DEBUG, "Checking to see if %s exists\n", exists_data.url->url);
+	http_curl_setopts(&curl_handle, headers, exists_data, url, cache);
+	switch_curl_easy_setopt(curl_handle, CURLOPT_FOLLOWLOCATION, 1);
+	switch_curl_easy_setopt(curl_handle, CURLOPT_MAXREDIRS, 10);
+	switch_curl_easy_setopt(curl_handle, CURLOPT_FAILONERROR, 1);
+	switch_curl_easy_setopt(curl_handle, CURLOPT_NOSIGNAL, 1);
+	if (headers) {
+		switch_curl_easy_setopt(curl_handle, CURLOPT_HTTPHEADER, headers);
+	}
+	switch_curl_easy_setopt(curl_handle, CURLOPT_NOBODY, 1);
+
+	curl_status = switch_curl_easy_perform(curl_handle);
+	switch_curl_easy_getinfo(curl_handle, CURLINFO_RESPONSE_CODE, &httpRes);
+	switch_curl_easy_cleanup(curl_handle);
+	close(exists_data.fd);
+	if (curl_status != CURLE_OK) {
+		url->size = 0; // nothing downloaded or download interrupted
+		status = SWITCH_STATUS_FALSE;
+		if (httpRes != 404) {
+			status = SWITCH_STATUS_GENERR;
+			switch_log_printf(SWITCH_CHANNEL_SESSION_LOG(session), SWITCH_LOG_ERROR, "Received curl error %d HTTP error code %ld on exists %s\n", curl_status, httpRes, url->url);
+		}
+		goto done;
+	}
+
+done:
+
+	if (headers) {
+		switch_curl_slist_free_all(headers);
+	}
+
+	return status;
+}
+
+/**
+ * Sets the options for a curl_handle
+ * @param cache the cache
+ * @param profile the HTTP profile
+ * @param url The cached URL entry
+ * @param session the (optional) session
+ * @return SWITCH_STATUS_SUCCESS if it exists 
+ * Will want to make this generic for head & delete & maybe get together but for now just trying to add functionality
+ */
+static switch_status_t http_curl_setopts(switch_CURL *curl_handle, switch_curl_slist_t *headers, http_get_data_t curl_data, cached_url_t *url, url_cache_t *cache)
+{
+	switch_status_t status = SWITCH_STATUS_SUCCESS;
+	/* find profile for domain */
+	switch_curl_easy_setopt(curl_handle, CURLOPT_FOLLOWLOCATION, 1);
+	switch_curl_easy_setopt(curl_handle, CURLOPT_MAXREDIRS, 10);
+	switch_curl_easy_setopt(curl_handle, CURLOPT_FAILONERROR, 1);
+	switch_curl_easy_setopt(curl_handle, CURLOPT_NOSIGNAL, 1);
+	if (headers) {
+		switch_curl_easy_setopt(curl_handle, CURLOPT_HTTPHEADER, headers);
+	}
+	switch_curl_easy_setopt(curl_handle, CURLOPT_URL, curl_data.url->url);
+	switch_curl_easy_setopt(curl_handle, CURLOPT_WRITEFUNCTION, get_file_callback);
+	switch_curl_easy_setopt(curl_handle, CURLOPT_WRITEDATA, (void *) &curl_data);
+	switch_curl_easy_setopt(curl_handle, CURLOPT_HEADERFUNCTION, get_header_callback);
```

Comment:
These should not be done inside this common function I think.  And it might not be limited to these...
The idea is to look at everywhere these switch_curl_easy_setopt are used in this file (which you created new one), then look at the similar options being set with all of them, and create that http_curl_setopts() function with those in them.

In this commit, you have only used that common function once.

What I would do is do a search for CURLOPT_ in the file, and look where they are use and find the one you can group together because they are similar and put them in that function.  It fine that there are easy_setopt remaining in the functions, there should just be a lot less of them, which will make clearer that those are needed for that particular function.

Reply:
So I think I had a bad merge/cherry-pick from working in a wrong branch that resulted in http_delete and http_exists being duplicated. I did use the http_curl_setopts in those functions in addition to http_get.
I looked at http_put and it was not similar enough to use http_curl_setopts for the following reasons.
It is making multiple requests inside a while loop. 
The curl handle is a CURL type not a  switch_CURL.
The options are not similar enough.

Regarding CURLOPT_WRITEFUNCTION & CURLOPT_WRITEDATA. I was not sure whether it should be included in the http_exists and http_delete requests because I do not care about the responses, however it looks like if I didn't use those than libcurl would use fwrite and I would need to set CURLOPT_WRITEDATA to a void file pointer but I figure if anyone after me wanted to expand the function to log the response on an error doing it this way should make that easier.

Reply:
Because when I looked at the code where it was used it was different at some place, I would keep outside of the common function for the moment, and someone could add them as parameters if needed in the future.  I'll say stick to all the parameters that are really assigned exactly the same way for your initial change.  We can move other at a latter date if needed but the frame work for it will be there.

Reply:
So I updated the code and added an empty_callback so it won't print to standard out for the http_delete and http_exist functions. Also I guess I forgot to reinstall when testing because I had a couple other errors which was causing crashes.

If you would like me to rebase this I can.

Reply:
You forgot : http_put as well